src/client/connstring.rs

```diff
@@ -0,0 +1,318 @@
+            }
+        },
+        false => partition(addr, "/")
+    };
+
+    if path_str.len() == 0 && host_str.contains("?") {
+        return Err("A '/' is required between the host list and any options.");
+    }
+
+    // Split on authentication and hosts
+    if host_str.contains("@") {
+        let (user_info, host_string) = rpartition(host_str, "@");
+        let (u,p) = try!(parse_user_info(user_info));
+        user = Some(u.to_owned());
+        password = Some(p.to_owned());
+        hosts = try!(split_hosts(host_string));
+    } else {
+        hosts = try!(split_hosts(host_str));
+    }
+
+    // Match connection type
+    ctype = match hosts.len() {
+        1 => ConnectionType::Master,
+        2 =>  ConnectionType::Pair,
+        n => {
+            if n > 2 {
+                ConnectionType::Set
+            } else {
+                ConnectionType::Invalid
+            }
+        }
+    };
+
+    let mut opts = "";
+
+    // Split on database name, collection, and options
+    if path_str.len() > 0 {
+        if path_str.starts_with("?") {
+            opts = &path_str[1..];
+        } else {
+            let (dbase, options) = partition(path_str, "?");
+            let (dbase_new, coll) = partition(dbase, ".");
+            database = Some(dbase_new.to_owned());
+            collection = Some(coll.to_owned());
+            opts = options;
+        }
+    }
+
+    // Collect options if any exist
+    if opts.len() > 0 {
+        options = Some(split_options(opts).unwrap());
+    }
+
+    Ok(ConnectionString {
+        ctype: ctype,
+        hosts: hosts,
+        string: Some(address.to_owned()),
+        user: user,
+        password: password,
+        database: database,
+        collection: collection,
+        options: options,
+    })
+}
+
+// Parse user information of the form user:password
+fn parse_user_info(user_info: &str) -> Result<(&str, &str), &str> {
+    let (user, password) = rpartition(user_info, ":");
+    if user_info.contains("@") || user.contains(":") {
+        return Err("':' or '@' characters in a username or password must be escaped according to RFC 2396.")
+    }
+    if user.len() == 0 {
+        return Err("The empty string is not a valid username.")
+    }
+    Ok((user, password))
+}
+
+// Parses a literal IPv6 literal host entity of the form [host] or [host]:port
+fn parse_ipv6_literal_host(entity: &str) -> Result<Host, &str> {
+    match entity.find("]") {
+        Some(_) => {
+            match entity.find("]:") {
+                Some(idx) => {
+                    let port = &entity[idx+2..];
+                    match port.parse::<u16>() {
+                        Ok(val) => Ok(Host::new(entity[1..idx].to_ascii_lowercase(), val)),
+                        Err(_) => Err("Port must be an integer."),
+                    }
+                },
+                None => Ok(Host::new(entity[1..].to_ascii_lowercase(), DEFAULT_PORT)),
+            }
+        }
```

Comment:
+1: not necessary to have the trailing comma but easier to move cases around (even if there are only two in this situation, generally better stylistically).